### ISSUE TITLE
No longer able to automatically create scenarios if they don't exist

### ISSUE THREAD

**Issue by atddkg** [Qytera-Gmbh/cypress-xray-plugin#358]
I had a few issues open a while back trying to get my head around using your plugin to use my automation framework as the single source of truth for creating/updating test tickets in Jira/Xray (assisted by @csvtuda ).

E.g. I would author the Cucumber scenarios in the automation suite and on the back of your plugin uploading the results, rely on Jira/Xray to create the test tickets if they don't exist or update them if they do.

This was working... for a bit. Now, for whatever reason, Jira/Xray is no longer automatically creating/updating test tickets on the back of your plugin uploading the results.

To get Jira/Xray to automatically create the test tickets, I tagged them with a never-existent dummy ticket number (TST-999999).

It used to not be able to find it and would then create the tickets for me. Now it stops at:

`"url":"https://censored.atlassian.net/rest/api/latest/search","data":"{\"fields\":[\"labels\"],\"jql\":\"issue in (TST-999999,TST-999999,TST-999999,TST-999999,TST-999999,TST-999999,TST-999999,TST-999999,TST-999999,TST-999999)\",\"startAt\":0}"},"code":"ERR_BAD_REQUEST","status":400},"response":{"errorMessages":["Issue does not exist or you do not have permission to see it."],"warningMessages":[]}}`

I realise this is almost certainly out of scope of your plugin but was just wondering if I could pick your brains about this?

Re: [https://docs.getxray.app/display/XRAY/Integrating+with+Testing+Frameworks](https://docs.getxray.app/display/XRAY/Integrating+with+Testing+Frameworks)

It states: "The results are mapped to the proper Test issues; if they don't exist then they will be created."

But it does then go on to state in the table:

"Create Tests from results
(create Tests whenever importing results)"

...

Cucumber:

"No*
(but there is an endpoint for importing Cucumber features, which will create Tests for the respective Scenarios/Scenario Outlines)"

I can't see if this page was changed but from memory, I'm pretty sure it's always said the above ... and yet I was able to do what it said I shouldn't be able to do.

For a second I was thinking I needed to implement a separate feature file upload but remembered your plugin does this:

`uploadFeatures: true,`

...assuming this is targeting [https://docs.getxray.app/display/XRAY/Importing+Cucumber+Tests+-+REST](https://docs.getxray.app/display/XRAY/Importing+Cucumber+Tests+-+REST) ?

This bit from the above link is no longer getting done - "create Test in that project and add a label with the relative path of the feature. The tags used in the scenario/scenario outline are also added as labels."

A bit confused...

Further information that may be relevant [Section: 'Using Git or other VCS as master'](https://docs.getxray.app/display/XRAY/Testing+using+Cypress+and+Cucumber+in+JavaScript)

Will read these docs to try to figure this out, apologies if this is completely out of scope of your plugin, just reaching out to soundboard this...

Will also post on Xray forum on the assumption your plugin is working fine.

**Comment by csvtuda** [comment]
Hey @atddkg,

thanks for the question. Don't worry, I think you are on the right track. Let me try to answer all your questions.

> It used to not be able to find it and would then create the tickets for me. Now it stops at:
> `"url":"https://censored.atlassian.net/rest/api/latest/search","data":"{\"fields\":[\"labels\"],\"jql\":\"issue in (TST-999999,TST-999999,TST-999999,TST-999999,TST-999999,TST-999999,TST-999999,TST-999999,TST-999999,TST-999999)\",\"startAt\":0}"},"code":"ERR_BAD_REQUEST","status":400},"response":{"errorMessages":["Issue does not exist or you do not have permission to see it."],"warningMessages":[]}}`
>
> I realise this is almost certainly out of scope of your plugin but was just wondering if I could pick your brains about this?

I could not find any changes to this endpoint in the latest Jira [changelogs](https://developer.atlassian.com/cloud/jira/platform/changelog/#). A little bit of trial and error makes me think that you might be updating _nothing but nonexistent_ features for the first time.

For example, the following Jira search request will return 400 Bad Request:

```json
{
  "jql": "issue in (CYP-999999999, CYP-000000000)",
  "startAt": 0,
  "fields": [
    "id", "summary", "labels"
  ]
}
```

Whereas the following one will return 200 Ok (CYP-885 exists in my project):

```json
{
  "jql": "issue in (CYP-999999999, CYP-000000000, CYP-885)",
  "startAt": 0,
  "fields": [
    "id", "summary", "labels"
  ]
}
```

So simply by including a valid issue in the search, Jira will return a successful response (including some warnings). Maybe this might help you already?

> ...assuming this is targeting https://docs.getxray.app/display/XRAY/Importing+Cucumber+Tests+-+REST?
> This bit from the above link is no longer getting done - "create Test in that project and add a label with the relative path of the feature. The tags used in the scenario/scenario outline are also added as labels."

You're right that the plugin uses this endpoint for importing/upstream synchronization of feature files. But it wouldn't help you, this is what the endpoint would return for a "dummied" feature file:

```gherkin
Feature: A cool story

    @TestName:CYP-999999
    Scenario: Scenario 1
        Given A Given
        When A When
        Then A Then

```

```json
{
    "errors": [
        "Error in file tagged.feature: Test with key CYP-999999 was not found!"
    ],
    "updatedOrCreatedTests": [],
    "updatedOrCreatedPreconditions": []
}
```

The only way you could leverage this endpoint is by:

1. importing an _untagged_ feature file
2. adding the corresponding tag afterwards

... which is something the plugin does not support (it does not upload untagged feature files).

---

From my perspective, it's not a plugin issue. But let me know what you think.

**Comment by atddkg** [comment]
"So simply by including a valid issue in the search (i.e. a correctly tagged feature file in your test run), Jira will return a successful response (including some warnings). Maybe this might help you already?"

Context: 1 feature file, 13 scenarios:

I created a dummy Test ticket in Jira and in my automation suite, tagged every new scenario with its issue ID (@TST-390). This resulted in some of the scenarios having Test tickets automatically created in Jira, but not all. 10/13 got automatically created.

importFeatureError.json...

`"_eventsCount":3}},"code":"ERR_BAD_REQUEST","status":400},"response":{"error":"Issue create failed!"}}`

I'm assuming `_eventsCount":3` refers to the 3 Test tickets that weren't created. Unfortunately, the error doesn't go into detail as to why their creation failed.

I repeated the above (deleting the automatically created Test tickets first) and the same thing happened - but it wasn't the same 3 scenarios that didn't get created.

Re-running the suite (without changing the tags of the 10 Test tickets that had already got created) resulted in the 3 scenarios having Test tickets automatically created for them.

Context: 11 feature files, 96 scenarios:

When I run and re-run my automation suite, it only added a random number of missing Test tickets each time:

Run 1: Created 66 Test tickets.

Run 2: Created 12 more Test tickets.

Run 3: Stopped creating Test tickets (8 remaining).

It's choking on 8 same-named scenarios in different feature files, e.g.

│ Cypress Xray Plugin │ WARNING │     - Error in file my-feature.feature: Test with summary "Unexpected request method returns correct response code and error message" and for project with id 11039 could not be univocally identified!

I gave these scenarios unique names and Run 4 created their Test tickets.

Out of curiosity, I deleted them all in Jira and ran the full suite again to see if it'd create them all in one go. It didn't.

Run 1: 61.

Run 2: 26 more.

Run 3: last 9.

At least I've got it doing what I want now, but the process is a bit janky.

There's probably a good reason for the "could not be univocally identified" error and non-creation of those scenarios as Test tickets, but the Scenario Outlines in question were more or less unique (with different examples in some instances).
